src/renderers/dom/client/MarkupMismatchError.js

```diff
@@ -0,0 +1,216 @@
+ * @param {String} the attribute value in the server markup
+ * @param {String} the attribute value in the client component tree.
+ */
+function throwAttributeChangedMismatchError(node, attr, serverValue, clientValue) {
+  throw new MarkupMismatchError(
+    `The value of the attribute '${attr}' differed.`,
+    node,
+    `${attr}=${serverValue}`,
+    `${attr}=${clientValue}`);
+}
+
+/**
+ * throw an error when html set with dangerouslySetInnerHTML has a
+ * different value in the server markup and client component tree.
+ * @param {DOMElement} the DOM element from the server markup that has the dangerouslySetInnerHTML
+ * @param {String} the dangerouslySetInnerHTML value in the server markup
+ * @param {String} the dangerouslySetInnerHTML value in the client component tree.
+ */
+function throwInnerHtmlMismatchError(node, serverValue, clientValue) {
+  throw new MarkupMismatchError(
+    'The value for dangerouslySetInnerHTML differed.',
+    node,
+    serverValue,
+    clientValue
+  );
+}
+
+/**
+ * throw an error when the text content of a node has a
+ * different value in the server markup and client component tree.
+ * @param {DOMElement} the DOM element from the server markup that has differing text
+ * @param {String} the text value in the server markup
+ * @param {String} the text value in the client component tree.
+ */
+function throwTextMismatchError(node, serverValue, clientValue) {
+  throw new MarkupMismatchError('Text content differed.', node, `'${serverValue}'`, `'${clientValue}'`);
+}
+
+/**
+ * throw an error when the component type (i.e. element, text, empty) of a node has a
+ * different value in the server markup and client component tree.
+ * @param {DOMNode|Array<DOMNode>} the DOM element or elements from the server
+ *   markup that is/are different on client
+ * @param {String} a user-readable description of the component in the client
+ *   component tree.
+ */
+function throwComponentTypeMismatchError(node, clientComponentDesc) {
+  var serverValue = '';
+
+  if (Array.isArray(node) && node.length === 3) {
+    serverValue = `A text node with text: '${abridgeContent(node[1].textContent)}'`;
+  } else if (Array.isArray(node) && node.length === 2) {
+    serverValue = 'A blank text node.';
+  } else if (node.nodeType === COMMENT_NODE_TYPE) {
+    serverValue = 'An empty (or null) node';
+  } else if (node.nodeType === ELEMENT_NODE_TYPE) {
+    serverValue = `A <${node.tagName.toLowerCase()}> element with text: '${abridgeContent(node.textContent)}'`;
+  } else {
+    serverValue = `An illegal DOM node with text: '${abridgeContent(node.textContent)}'`;
+  }
+
+  throw new MarkupMismatchError(
+    'The type of component differed.',
+    Array.isArray(node) ? node[0] : node,
+    serverValue,
+    clientComponentDesc);
+}
+
+/**
+ * truncate text utility function.
+ * @private
+ * @param {String} text to truncate
+ * @param {Number?} maximum length
+ */
+function abridgeContent(text, maxLength = 30) {
```

Comment:
I wasn't sure if there was a utility function that already did this; please let me know if there is.